tools/docs/configuration.md
# Configuration

This service exposes several configuration options that are controlled via environment variables. There are several available that can be used to configure how it behaves:

| Environment Variable | Description                                                                                                                                                                                                                                                                                                              |
|----------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| SKIP_MIGRATIONS      | Whether to skip executing database migrations on the outbox table. By default, this is false, and should usually be OK for most cases. If you skip migrations then you will be at risk of updates to this relay service not behaving as expected. |
| DB_HOST              | The database host where the outbox table resides.                                                                                                                                                                                                                                                                        |
| DB_PORT              | Database port.                                                                                                                                                                                                                                                                                                           |
| DB_USER              | Database user.                                                                                                                                                                                                                                                                                                           |
| DB_PASS              | Database password.                                                                                                                                                                                                                                                                                                       |
| DB_NAME              | Database name.                                                                                                                                                                                                                                                                                                           |
| DB_DRIVER            | The type of database driver to use, options are either "mysql" or "postgres".                                                                                                                                                                                                                                            |
| KAFKA_HOST           | The Kafka host, should be comma separated when there are multiple Kafka brokers, e.g. "kafka1:9092,kafka2:9092"                                                                                                                                                                                                          |
| TLS_ENABLE           | Whether to enable TLS when communicating with Kafka and the database. We recommend enabling this if your database and Kafka cluster support it. Defaults to false.                                                                                                                                                       |
| TLS_SKIP_VERIFY_PEER | Whether to skip peer verification when connecting over TLS. Defaults to false.                                                                                                                                                                                                                                           |
| WRITE_CONCURRENCY    | The number of concurrent workers used to push data to Kafka. Defaults to 1. You should only need to increase this if the throughput of messages to the outbox is extremely high.                                                                                                                                         |
| POLL_FREQUENCY_MS    | How frequently, in milliseconds, to poll the outbox table for new messages after an error is encountered, or no events were found on the last poll. Defaults to 500.                                                                                                                                                                                                                             |
| BATCH_SIZE           | The maximum number of messages to grab from the outbox table for each poll operation. Defaults to 250.                                                                                                                                                                                                                   |
| POLLING_DISABLED     | When set to true, the outbox relay will not poll for messages and will not attempt to connect to Kafka. This is useful when you want to run the outbox relay in your local stack to facilitate development. Defaults to false.                                                                                   |


config/config.go
package config

import (
	"encoding/json"
	"fmt"
	"net/url"
	"time"

	"inviqa/kafka-outbox-relay/log"

	"github.com/alexflint/go-arg"
)

const (
	MySQL    DbDriver = "mysql"
	Postgres DbDriver = "postgres"

	defaultPublishAttempts = 3
	outboxTable            = "kafka_outbox"
)

type DbDriver string

var supportedDbTypes = map[DbDriver]bool{
	Postgres: true,
	MySQL:    true,
}

type Config struct {
	PollingDisabled      bool     `arg:"--polling-disabled,env:POLLING_DISABLED"`
	SkipMigrations       bool     `arg:"--skip-migrations,env:SKIP_MIGRATIONS"`
	DBHost               string   `arg:"--db-host,env:DB_HOST,required"`
	DBPort               uint32   `arg:"--db-port,env:DB_PORT,required"`
	DBUser               string   `arg:"--db-user,env:DB_USER,required"`
	DBPass               string   `arg:"--db-pass,env:DB_PASS,required"`
	DBName               string   `arg:"--db-name,env:DB_NAME,required"`
	DBDriver             DbDriver `arg:"--db-driver,env:DB_DRIVER,required"`
	DBOutboxTable        string
	KafkaHost            []string `arg:"--kafka-host,env:KAFKA_HOST"`
	KafkaPublishAttempts int      `arg:"--kafka-publish-attempts,env:KAFKA_PUBLISH_ATTEMPTS"`
	TLSEnable            bool     `arg:"--kafka-tls,env:TLS_ENABLE"`
	TLSSkipVerifyPeer    bool     `arg:"--kafka-tls-verify-peer,env:TLS_SKIP_VERIFY_PEER"`
	WriteConcurrency     int      `arg:"--write-concurrency,env:WRITE_CONCURRENCY"`
	PollFrequencyMs      int      `arg:"--poll-frequency-ms,env:POLL_FREQUENCY_MS"`
	RunCleanup           bool     `arg:"--cleanup,env:RUN_CLEANUP"`
	RunOptimize          bool     `arg:"--optimize,env:RUN_OPTIMIZE"`
	SidecarProxyUrl      string   `arg:"--sidecar-proxy-url,env:SIDECAR_PROXY_URL"`
	BatchSize            int      `arg:"--batch-size,env:BATCH_SIZE"`
}

func NewConfig() (*Config, error) {
	c := &Config{
		KafkaPublishAttempts: defaultPublishAttempts,
		DBOutboxTable:        outboxTable,
		WriteConcurrency:     1,
		PollFrequencyMs:      500,
		BatchSize:            250,
	}
	arg.MustParse(c)

	if !supportedDbTypes[c.DBDriver] {
		return nil, fmt.Errorf("the DB_DRIVER provided (%s) is not supported", c.DBDriver)
	}

	return c, nil
}

func (c *Config) GetPollIntervalDurationInMs() time.Duration {
	return time.Duration(c.PollFrequencyMs) * time.Millisecond
}

func (c *Config) GetDSN() string {
	switch c.DBDriver {
	case MySQL:
		tls := "false"
		if c.TLSEnable {
			if c.TLSSkipVerifyPeer {
				tls = "skip-verify"
			} else {
				tls = "true"
			}
		}
		return fmt.Sprintf("%s:%s@tcp(%s:%d)/%s?parseTime=true&tls=%s&multiStatements=true", c.DBUser, c.DBPass, c.DBHost, c.DBPort, c.DBName, tls)
	case Postgres:
		sslMode := "disable"
		if c.TLSEnable {
			if c.TLSSkipVerifyPeer {
				sslMode = "require"
			} else {
				sslMode = "verify-full"
			}
		}
		return fmt.Sprintf("%s://%s@%s:%d/%s?sslmode=%s", c.DBDriver, url.UserPassword(c.DBUser, c.DBPass), c.DBHost, c.DBPort, c.DBName, sslMode)
	default:
		log.Logger.Fatalf("the DB driver configured (%s) is not supported", c.DBDriver)
		return ""
	}
}

func (c *Config) GetDependencySystemAddresses() []string {
	return c.KafkaHost
}

func (c Config) MarshalJSON() ([]byte, error) {
	return json.Marshal(map[string]interface{}{
		"PollingDisabled":      c.PollingDisabled,
		"SkipMigrations":       c.SkipMigrations,
		"DBHost":               c.DBHost,
		"DBPort":               c.DBPort,
		"DBUser":               c.DBUser,
		"DBPass":               "xxxxx",
		"DBName":               c.DBName,
		"DBDriver":             c.DBDriver,
		"KafkaHost":            c.KafkaHost,
		"KafkaPublishAttempts": c.KafkaPublishAttempts,
		"TLSEnable":            c.TLSEnable,
		"TLSSkipVerifyPeer":    c.TLSSkipVerifyPeer,
		"WriteConcurrency":     c.WriteConcurrency,
		"PollFrequencyMs":      c.PollFrequencyMs,
		"RunCleanup":           c.RunCleanup,
		"RunOptimize":          c.RunOptimize,
		"SidecarProxyUrl":      c.SidecarProxyUrl,
		"BatchSize":            c.BatchSize,
	})
}

func (d DbDriver) MySQL() bool {
	return d == MySQL
}

func (d DbDriver) Postgres() bool {
	return d == Postgres
}

func (d DbDriver) String() string {
	return string(d)
}
